[fp-rs] Implemented Fast Pair Seeker v1 for Windows in Rust.

TheShepord opened the pull request:
## Summary
This change implements Fast Pair Seeker v1 for Windows in Rust. This involved adding the following functionality to the existing scanning implementation:
* Address API between bluetooth module and seeker.
* Classic device pairing.
* Retrieving service data from advertisement packet and filtering FP advertisements.
* Device selection CLI.

## How did you test this change?
`cargo run`, then pairing with a FP v1 device (LE_WH-1000XM3).

TheShepord commented:
@suetfei @anayw2001 @marshallpierce

anayw2001 commented on fastpair/rust/src/bluetooth/common/address.rs:
```diff
@@ -0,0 +1,101 @@
+    kind: BleAddressKind,
+}
+
+/// Struct representing a 48-bit BLE classic address.
```
BT classic*

anayw2001 commented on fastpair/rust/src/bluetooth/windows/address.rs:
```diff
@@ -0,0 +1,49 @@
+            )),
+            _ => Err(anyhow::anyhow!(
+                "Attempting to construct `BleAddressKind` with device \
+                advertising invalid address type ."
```
maybe debug log the type here?

marshallpierce replied:
Or just include the type in the error

anayw2001 commented on fastpair/rust/src/bluetooth/windows/device.rs:
```diff
@@ -13,34 +13,59 @@
+            // Enum describing the type of address (public, random, unspecified).
+            // https://learn.microsoft.com/en-us/uwp/api/windows.devices.bluetooth.bluetoothaddresstype?view=winrt-22621
+            BluetoothAddressType,
+            // TODO
```
is this a doc todo?

suetfei replied:
TODO(b/xxxx): ...

marshallpierce commented on fastpair/rust/src/bluetooth/common/data.rs:
```diff
@@ -0,0 +1,36 @@
+        ServiceData { uuid, data }
+    }
+
+    pub fn get_uuid(&self) -> Uuid {
```
getters don't have a get_ prefix typically

anayw2001 commented on fastpair/rust/src/bluetooth/windows/device.rs:
```diff
@@ -49,6 +74,97 @@ impl Device for BleDevice {
+impl ClassicDevice {
+    /// `ClassicDevice` constructor.
+    pub async fn new(addr: ClassicAddress) -> Result<Self, anyhow::Error> {
+        let raw_addr = u64::from(addr);
+
+        let inner = BluetoothDevice::FromBluetoothAddressAsync(
+            raw_addr,
+        )?
```
can we inline this

anayw2001 replied:
e.g.
`let inner = BluetoothDevice::FromBluetoothAddressAsync(u64::from(addr))?.await?;`

marshallpierce commented on fastpair/rust/src/bluetooth/common/device.rs:
```diff
@@ -0,0 +1,35 @@
+    /// Attempt pairing with the peripheral device.
+    async fn pair(&self) -> Result<(), anyhow::Error>;
+
+    fn get_service_data(&self) -> &Vec<ServiceData<u16>>;
```
I'm not sure it makes sense for a Device to have service data. Advertisements might, but a device isn't an advertisement

anayw2001 replied:
should we have an advertisement struct and provide a fn in device to get the advertisement the device was built from?

TheShepord replied:
@anayw2001 this is what I was thinking as well, will take a bit of refactoring so I'll check it in as a separate commit.

anayw2001 commented on fastpair/rust/src/bluetooth/windows/device.rs:
```diff
@@ -49,6 +74,97 @@ impl Device for BleDevice {
+#[async_trait]
+impl Device for ClassicDevice {
+    fn name(&self) -> Result<String, anyhow::Error> {
+        Ok(self.inner.Name()?.to_string_lossy())
```
can we use normal to_string() here? i feel like windows requires the adapter names to be utf-8 encoded.

anayw2001 commented on fastpair/rust/src/bluetooth/windows/device.rs:
```diff
@@ -49,6 +74,97 @@ impl Device for BleDevice {
+            println!("Device already paired");
+            Ok(())
+        } else if !pair_info.CanPair()? {
+            println!("Device can't pair");
```
can we make these printlns debug logs?

anayw2001 commented on fastpair/rust/src/bluetooth/windows/adapter.rs:
```diff
@@ -55,18 +54,21 @@ use windows::{
+
+/// Struct holding the necessary fields for listening to and handling incoming
+/// BLE advertisements.
+struct AdListener {
```
advlistener, adlistener is not quite clear enough

anayw2001 commented on fastpair/rust/src/bluetooth/windows/adapter.rs:
```diff
@@ -160,60 +162,51 @@ impl Adapter for BleAdapter {
+
                         match BleDevice::new(addr).await {
-                            Ok(device) => Some(device),
+                            Ok(device) => break Ok(device),
```
is the break needed here?

anayw2001 commented on fastpair/rust/src/bluetooth/common/data.rs:
```diff
@@ -0,0 +1,36 @@
+    ServiceData16BitUUid = 0x16,
+}
+
+pub struct ServiceData<Uuid: Copy> {
```
service data is technically a map of uuid -> data, so that would be more representative of the advertisement.

anayw2001 commented on fastpair/rust/src/bluetooth/common/data.rs:
```diff
@@ -0,0 +1,36 @@
+// limitations under the License.
+
+pub enum BleDataSection {
+    ServiceData16BitUUid = 0x16,
```
this may be a slight misunderstanding with regards to the name of this struct, see https://stackoverflow.com/questions/57683171/where-to-find-service-data-uuids

anayw2001 commented on fastpair/rust/src/bluetooth/windows/adapter.rs:
```diff
@@ -71,6 +76,44 @@ pub struct BleAdapter {
+
+        let first_byte = data_reader.ReadByte()?;
+        let second_byte = data_reader.ReadByte()?;
+        let uuid = ((first_byte as u16) << 8) | (second_byte as u16);
```
datareader has a readint16 fn, should we just use that?

anayw2001 commented on fastpair/rust/src/bluetooth/windows/adapter.rs:
```diff
@@ -71,6 +76,44 @@ pub struct BleAdapter {
+        let mut data = Vec::new();
+
+        for _ in 0..unconsumed_buffer_len {
+            data.push(data_reader.ReadByte()?);
```
can call DataReader::ReadBuffer(len: u32) directly.

anayw2001 commented on fastpair/rust/src/main.rs:
```diff
@@ -24,36 +24,37 @@ fn main() -> Result<(), anyhow::Error> {
+                    // shared cross-platform behavior.
+                    let classic_addr = match addr {
+                        Address::Ble(ble) => ClassicAddress::try_from(ble),
+                        Address::Classic(_) => unreachable!(
```
panic here, better style imo

anayw2001 commented on fastpair/rust/src/bluetooth/common/device.rs:
```diff
@@ -21,7 +21,7 @@ use crate::bluetooth::common::ServiceData;
 /// such as pairing.
 #[async_trait]
-pub trait Device: Sized {
+pub trait Device: Sized + Send + Sync {
```
correct me if i'm wrong here but this seems... strange.

anayw2001 replied:
@marshallpierce

marshallpierce replied:
Hard to say... You'd need this if Device is intended to be shared across threads, but hard to say in isolation if that's actually reasonable.

TheShepord commented:
Implemented all closed changes. Added library error types. @anayw2001 @marshallpierce 

TODO: decouple service_data from Device into Advertisement abstraction. Devices can then be subsequently created from Advertisements.

TheShepord commented:
Decoupled service data from device by creating separate advertisement struct. Ready for review @anayw2001, @marshallpierce .

anayw2001 commented on fastpair/rust/Cargo.toml:
```diff
@@ -20,11 +20,12 @@ edition = "2021"
 cfg-if = "1.0.0"
 async-trait = "0.1"
+thiserror = "1.0.43"
+anyhow = "1.0.72"
```
can we drop anyhow now?

anayw2001 commented on fastpair/rust/src/bluetooth/windows/error.rs:
```diff
@@ -0,0 +1,21 @@
+
+use crate::bluetooth::common::BluetoothError;
+
+impl From<windows::core::Error> for BluetoothError {
```
this might be better suited to just stick in mod.rs since it's just one fn

anayw2001 commented on fastpair/rust/src/bluetooth/common/address.rs:
```diff
@@ -0,0 +1,100 @@
+}
+
+impl TryFrom<BleAddress> for ClassicAddress {
+    // TODO proper error handling b/
```
can we specify an error type for this

anayw2001 commented on fastpair/rust/src/main.rs:
```diff
@@ -16,18 +16,48 @@ use futures::executor;
+                // cross-platform behavior.
+                let classic_addr = match addr {
+                    Address::Ble(ble) => ClassicAddress::try_from(ble),
+                    Address::Classic(_) => unreachable!(
```
we should probably panic here instead of marking it unreachable, it'll be the same either way in terms of runtime but having unreachable in a potentially reachable condition seems wrong.

anayw2001 commented on fastpair/rust/src/bluetooth/windows/adapter.rs:
```diff
@@ -51,11 +51,15 @@ use windows::{
     // https://learn.microsoft.com/en-us/uwp/api/windows.foundation.typedeventhandler-2?view=winrt-22621
     Foundation::TypedEventHandler,
+
+    // TODO
```
?

anayw2001 commented on fastpair/rust/src/bluetooth/windows/device.rs:
```diff
@@ -107,7 +109,7 @@ impl ClassicDevice {
     pub async fn new(addr: ClassicAddress) -> Result<Self, BluetoothError> {
         let raw_addr = u64::from(addr);
-
+        
```
whitespace

anayw2001 commented on fastpair/rust/src/bluetooth/windows/advertisement.rs:
```diff
@@ -0,0 +1,108 @@
+        let unconsumed_buffer_len = data_reader.UnconsumedBufferLength()?;
+        let mut data = Vec::new();
+        for _ in 0..unconsumed_buffer_len {
+            data.push(data_reader.ReadByte()?);
```
DataBuffer has a read method you can use to specify # of bytes
